bin/compare-factory-config.py

```diff
@@ -0,0 +1,180 @@
+    | root
+    | --ResourceGroup
+    | ----Facility
+    | ----Site
+    | ----Resources
+    | --------Resource
+
+    Code below is for parsing xml files.
+    topologyTree = ET.parse("resource_topology.xml")
+    topologyRoot = topologyTree.getroot()
+    """
+    response = urllib.request.urlopen(
+        "https://topology.opensciencegrid.org/rgsummary/xml?gridtype=on&gridtype_1=on&active=on&active_value=1&service=on&service_1=on")
+    topologyPage = response.read()
+    topologyRoot = ET.fromstring(topologyPage)
+
+    for child in topologyRoot.findall('ResourceGroup'):
+        # adding resourceGroup Name attribute to a set
+        name = child.find('GroupName')
+        if treeDump:
+            print("| " + name.text)
+        topologyDB['resourceGroups'].add(name.text)
+
+        for facility in child.findall('Facility'):
+            facilityName = facility.find('Name')
+            if treeDump:
+                print("| ---- " + facilityName.text)
+            topologyDB['facilities'].add(facilityName.text)
+        for site in child.findall('Site'):
+            siteName = site.find('Name')
+            if treeDump:
+                print("| ---- " + siteName.text)
+            topologyDB['sites'].add(siteName.text)
+        for resources in child.findall('Resources'):
+            for resource in resources.findall('Resource'):
+                resourceName = resource.find('Name')
+                if treeDump:
+                    print("| >>>> " + resourceName.text)
+                topologyDB['resources'].add(resourceName.text)
+
+
+def getGfactoryData(gfactoryDB, file):
+    """
+    Code below is for parsing xml URLs.
+    # response = urllib.request.urlopen(xml)
+    # gfactoryPage = response.read()
+    # root = ET.fromstring(gfactoryPage)
+    """
+    if (file.endswith('xml')):
+        tree = ET.parse(file)
+        root = tree.getroot()
+        # insert Names in Topology database into
+        for entry in root.findall('entries/entry'):
+            for attr in entry.findall('attrs/attr'):
+                if attr.get('name') == 'GLIDEIN_ResourceName':
+                    if treeDump:
+                        print(attr.get('value'))
+                    # gfactory structure: {GLIDEIN_ResourceName: entry name, ...}
+                    gfactoryDB[attr.get('value')] = entry.get('name')
+    else:
+        with open(file, 'r') as stream:
+            try:
+                data = yaml.safe_load(stream)
+                for resource in data.values():
+                    for entry in resource.values():
+                        for entryName, attr in entry.items():
+                            for tag, tagValue in attr.items():
+                                if (tag == 'attrs' and 'GLIDEIN_ResourceName' in list(tagValue.keys())):
+                                    gfactoryDB[tagValue['GLIDEIN_ResourceName']
+                                               ['value']] = entryName
```

Comment:
```suggestion
                for resource in data.values():
                    for entry in resource.values():
                        for entry_name, config in entry.items():
                            resource_name = config['attrs']['GLIDEIN_ResourceName']
                            gfactory_db[resource_name] = entry_name
```

This is starting to get a little too nested for my tastes. Something like the above is easier to follow and if you wrap this in a try/except with a `continue`, then it should skip malformed entries, which is more Pythonic (better to ask for forgiveness than permission).  Additionally, you should take this looping out of the `with` block and exception handling for an issue loading the YAML file.

Reply:
That looks way better